python/ray/serve/application.py

```diff
@@ -0,0 +1,190 @@
+from ray.serve.utils import logger
+
+
+class Application:
+    def __init__(self, deployments: List[Deployment]):
+        self._deployments = dict()
+        for idx, d in enumerate(deployments):
+            if isinstance(d, Deployment):
+                self._deployments[d.name] = d
+            else:
+                raise TypeError(
+                    f"Got object of type {type(d)} at index "
+                    f"{idx}. Expected only Deployment objects to "
+                    "be passed in."
+                )
+
+    def add_deployment(self, deployment: Deployment):
+        """Add deployment to the list. Validate name uniqueness."""
+
+        if not isinstance(deployment, Deployment):
+            raise TypeError(f"Got {type(deployment)}. Expected deployment.")
+
+        self._deployments[deployment.name] = deployment
+
+    def deploy(self, blocking: bool = True):
+        """Async deploy."""
+
+        if len(self._deployments) == 0:
+            return
+
+        parameter_group = []
+
+        for deployment in self._deployments:
+            if not isinstance(deployment, Deployment):
+                raise TypeError(
+                    f"deploy_group only accepts Deployments, but got unexpected "
+                    f"type {type(deployment)}."
+                )
+
+            deployment_parameters = {
+                "name": deployment._name,
+                "func_or_class": deployment._func_or_class,
+                "init_args": deployment.init_args,
+                "init_kwargs": deployment.init_kwargs,
+                "ray_actor_options": deployment._ray_actor_options,
+                "config": deployment._config,
+                "version": deployment._version,
+                "prev_version": deployment._prev_version,
+                "route_prefix": deployment.route_prefix,
+                "url": deployment.url,
+            }
+
+            parameter_group.append(deployment_parameters)
+
+        return _get_global_client().deploy_group(parameter_group, _blocking=blocking)
+
+    @staticmethod
+    def get_statuses() -> Dict[str, DeploymentStatus]:
+        """Get current status of all deployments."""
+
+        return _get_global_client().get_deployment_statuses()
+
+    def run(
+        self,
+        runtime_env_updates=None,
+        cluster_address="auto",
+        dashboard_address="http://localhost:8265",
+        logger=logger,
+    ):
+        """Blocking run."""
+
+        runtime_env_updates = runtime_env_updates or {}
+
+        ray.init(address=cluster_address, namespace="serve")
+        serve.start(detached=True)
+        ServeSubmissionClient(dashboard_address)._upload_working_dir_if_needed(
+            runtime_env_updates
+        )
+
+        for deployment in self._deployments:
+            self._configure_runtime_env(deployment, runtime_env_updates)
+        self.deploy()
+
+        logger.info("\nDeployed successfully!\n")
+
+        while True:
```

Comment:
hmm why do we need to bundle app deploy with a while loop for deployment status polling ? For our end users who might be using their internal tooling to manage serve deployments this could be a bit too limiting. How about split this part into something like `poll` or `watch` that user needs to explicitly call into, rather than looping here forever ? By default we can still surface all deployment logs from serve via `self.deploy()`

Reply:
`run` and `deploy` are different. `deploy` submits deployments to the ray cluster and then returns. `run` submits deployments and then keeps looping until manually killed. When killed, `run` tears down the deployments it submitted, and if there are no deployments left, it tears down the rest of Serve (e.g. the controller). This way, users can write a Python script with `run`, and manually terminate it. It's similar to the cli's `serve run`.

Reply:
Ah i see, this makes sense if we tear down deployments upon exiting. Then my inputs here will be docstring + examples